File: ncarg2d/man/conpack/cpscae.m
.TH CPSCAE 3NCARG "March 1993" UNIX "NCAR GRAPHICS"
.na
.nh
.SH NAME
CPSCAE -
Called repeatedly by the routine CPCICA as it executes. Updates a particular
element of a user's cell array.  The default version of CPSCAE may be
replaced by the user with a version that does something different.
.SH SYNOPSIS
 CALL CPSCAE (ICRA, ICA1, ICAM, ICAN, XCPF, YCPF, XCQF, 
.br
+ YCQF, IND1, IND2, ICAF, IAID)
.SH DESCRIPTION 
The first eight arguments of CPSCAE describe the user's 
cell array and are taken directly from the user's call to 
CPCICA:
.IP ICRA 12
(INTEGER array, dimensioned ICA1 by "n", where "n" is 
greater than or equal to the value of the argument ICAN, 
input/output) is the user's cell array. Generally speaking, 
a user version of CPSCAE should only reset the value of the 
single element ICRA(IND1,IND2).
.IP ICA1 12
(INTEGER, input) is the first dimension of the FORTRAN 
array ICRA, which contains the user's cell array.
.IP ICAM 12
(INTEGER, input) is the first dimension of the user's 
cell array.
.IP ICAN 12
(INTEGER, input) is the second dimension of the user's 
cell array.
.IP "XCPF and YCPF" 12
(REAL, input) are the coordinates, in the 
fractional coordinate system, of a point P. P is the point 
at that corner of the rectangular area into which the cell 
array maps that corresponds to the cell (1,1).
.IP "XCQF and YCQF" 12
(REAL, input) are the coordinates, in the 
fractional coordinate system, of a point Q. Q is the point 
at that corner of the rectangular area into which the cell 
array maps that corresponds to the cell (ICAM,ICAN).
.IP "IND1 and IND2" 12
(INTEGER, input) are the indices of an 
element in the cell array that is to be reset.
.IP ICAF 12
(INTEGER, input) is the value of the internal 
parameter 'CAF', which will be a negative integer (because, 
when 'CAF' is greater than or equal to zero, the routine 
CPSCAE is not called). Different values of 'CAF' may be 
used in a user version of CPSCAE to select different 
schemes for picking color indices.
.IP IAID 12
(INTEGER, input) is the value of the area identifier 
associated with the midpoint of the cell (IND1,IND2). The 
four possibilities are as described for the routine CPCICA.
.SH USAGE
CPSCAE is not called by the user. When a user calls the
routine CPCICA and the internal parameter 'CAF' is
negative, CPSCAE is called by CPCICA exactly once for each
cell in the cell array. The default version of CPSCAE just
sets the cell's color index equal to the area identifier
associated with the cell, if that area identifier is
greater than or equal to zero; otherwise, it does nothing.
A user-supplied version of CPSCAE may do something more
complicated; for example, it may take into account both the
value of the area identifier associated with the cell and
the current value of the color index for the cell in
determining the new value of the color index for the cell.
.SH ACCESS
To use CPSCAE, load the NCAR Graphics libraries ncarg, ncarg_gks,
and ncarg_c, preferably in that order.  
.SH SEE ALSO
Online:
conpack,
cpback, cpchcf, cpchcl, cpchhl, cpchil, cpchll, cpcica, cpclam, cpcldm,
cpcldr, cpcltr, cpcnrc, cpdrpl, cpezct, cpgetc, cpgeti, cpgetr, cplbam,
cplbdr, cpmpxy, cpmviw, cpmvrw, cppkcl, cppklb, cprect, cprset, cpsetc,
cpseti, cpsetr, cpsps1, cpsps2, ncarg_cbind
.SH COPYRIGHT
Copyright (C) 1987-2009
.br
University Corporation for Atmospheric Research
.br

The use of this Software is governed by a License Agreement.

File: ncarg2d/src/libncarg/conpack/cpscae.f
C
C $Id: cpscae.f,v 1.6 2000-07-12 16:22:35 haley Exp $
C                                                                      
C                Copyright (C)  2000
C        University Corporation for Atmospheric Research
C                All Rights Reserved
C
C This file is free software; you can redistribute it and/or modify
C it under the terms of the GNU Lesser General Public License as
C published by the Free Software Foundation; either version 2.1 of the
C License, or (at your option) any later version.
C
C This software is distributed in the hope that it will be useful, but
C WITHOUT ANY WARRANTY; without even the implied warranty of
C MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
C Lesser General Public License for more details.
C
C You should have received a copy of the GNU Lesser General Public
C License along with this software; if not, write to the Free Software
C Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
C USA.
C
      SUBROUTINE CPSCAE (ICRA,ICA1,ICAM,ICAN,XCPF,YCPF,XCQF,YCQF,
     +                                       IND1,IND2,ICAF,IAID)
      DIMENSION ICRA(ICA1,*)
C
C This routine is called by CPCICA when the internal parameter 'CAF' is
C given a negative value.  Each call is intended to update a particular
C element in the user's cell array.  The arguments are as follows:
C
C ICRA is the user's cell array.
C
C ICA1 is the first dimension of the FORTRAN array ICRA.
C
C ICAM and ICAN are the first and second dimensions of the cell array
C stored in ICRA.
C
C (XCPF,YCPF) is the point at that corner of the rectangular area
C into which the cell array maps that corresponds to the cell (1,1).
C The coordinates are given in the fractional coordinate system (unlike
C what is required in a call to GCA, in which the coordinates of the
C point P are in the world coordinate system).
C
C (XCQF,YCQF) is the point at that corner of the rectangular area into
C which the cell array maps that corresponds to the cell (ICAM,ICAN).
C The coordinates are given in the fractional coordinate system (unlike
C what is required in a call to GCA, in which the coordinates of the
C point Q are in the world coordinate system).
C
C IND1 is the 1st index of the cell that is to be updated.
C
C IND2 is the 2nd index of the cell that is to be updated.
C
C ICAF is the current value of the internal parameter 'CAF'.  This
C value will always be an integer which is less than zero (because
C when 'CAF' is zero or greater, this routine is not called).
C
C IAID is the area identifier associated with the cell.  It will have
C been given one of the values from the internal parameter array 'AIA'
C (the one for 'PAI' = -2 if the cell lies in a special-value area, the
C one for 'PAI' = -1 if the cell lies off the data grid, or the one for
C some value of 'PAI' between 1 and 'NCL' if the cell lies on the data
C grid).  The value zero may occur if the cell falls in a special-value
C area and the value of 'AIA' for 'PAI' = -2 is 0 or if the cell lies
C off the data grid and the value of 'AIA' for 'PAI' = -1 is 0, or
C if the cell falls on the data grid, but no contour level below the
C cell has a non-zero 'AIA' and no contour level above the cell has a
C non-zero 'AIB'.  Note that, if the values of 'AIA' for 'PAI' = -1
C and -2 are given non-zero values, IAID can only be given a zero
C value in one way.
C
C The default behavior of CPSCAE is as follows:  If the area identifier
C is non-negative, it is treated as a color index, to be stored in the
C appropriate cell in the cell array, but if the area identifier is
C negative, the cell array is simply not changed.  The user may supply
C a version of CPSCAE that does something different; it may simply map
C the area identifiers into color indices or it may somehow modify the
C existing cell array element to incorporate the information provided
C by the area identifier.
C
      IF (IAID.GE.0) ICRA(IND1,IND2)=IAID
C
      RETURN
C
      END
